draft-ietf-privacypass-integration.md

```diff
@@ -0,0 +1,147 @@
+
+stand_alone: yes
+pi: [toc, sortrefs, symrefs]
+
+author:
+ -
+    ins: A. Davidson
+    name: Alex Davidson
+    org: Brave Software
+    city: Lisbon
+    country: Portugal
+    email: alex.davidson92@gmail.com
+
+normative:
+  RFC2119:
+  RFC8446:
+  I-D.irtf-cfrg-voprf:
+  I-D.ietf-privacypass-architecture:
+  I-D.ietf-privacypass-http-api:
+informative:
+  RFC7049:
+  RFC7159:
+
+--- abstract
+
+This document specifies the necessary integration for building the
+Privacy Pass protocol, supporting symmetric verification and public
+metadata, on top of an existing oblivious pseudorandom function
+protocol.
+
+--- middle
+
+# Introduction
+
+The Privacy Pass protocol provides a privacy-preserving authorization
+mechanism. In essence, the protocol allows clients to provide
+cryptographic tokens that prove nothing other than that they have been
+authorized by a given server in the past
+{{I-D.ietf-privacypass-architecture}}.
+
+This document provides the necessary integration for building the
+authorization framework, based on existing constructions of oblivious
+pseudorandom function protocols {{I-D.irtf-cfrg-voprf}}. Moreover, we
+show how this integration allows `public` metadata to be introduced to
+the protocol, that is agreed by both clients and servers.
+
+This document DOES NOT cover the architectural framework required for
+running and maintaining the Privacy Pass protocol in the Internet
+setting. In addition, it DOES NOT cover the choices that are necessary
+for ensuring that client privacy leaks do not occur. Both of these
+considerations are covered in {{I-D.ietf-privacypass-architecture}}. In
+addition, considerations of how to embed the protocol interactions in
+the HTTP setting are considered in {{I-D.ietf-privacypass-http-api}}.
+
+# Terminology
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in {{RFC2119}}.
+
+The following terms are used throughout this document.
+
+- Client: An entity that provides authorization tokens to services
+  across the Internet, in return for authorization.
+- Server: A service (also known as an Issuer) that issues Privacy Pass
+  tokens to clients.
+- Key: The secret key used by the server for authorizing client data.
+- Blinding/unblinding: A pair of cryptographic operations that allow
+  randomly distributing and revealing some underlying `data`, whilst
+  preserving the ability to generate valid signatures on `data` itself.
+
+We assume that all protocol messages are encoded into raw byte format
+before being sent across the wire.
+
+# Privacy Pass flow {#protocol-flow}
+
+There are three phases in the protocol: the initialization phase, the
+issuance phase, and the redemption phase.
+
+In the initialization phase, the server samples a keypair and publishes
+a ``public key`` (or key commitment) to a public repository.
+
+In the issuance phase:
```

Comment:
I would spell this differently. For example, I might literally frame it in terms of the POPRF protocol, like so:

```
The issuance phase proceeds as follows. Let `info` be the agreed upon metadata between client
and server, and let `config` be the server's chosen key configuration.

First, a client configures its verifiable context using `config`:

~~~
client_context = SetupVerifiableClient(config.suite, config.public_key)
~~~

Likewise, the server creates its own context using `config` and the corresponding private key `key`:

~~~
server_context = SetupVerifiableServer(config.suite, key, config.public_key)
~~~

The client then creates an issuance request as follows:

~~~
nonce = random(32)
blind, blindedElement = context. Blind(nonce)
~~~

The client then sends blindedElement to the server. The server, upon receipt, evaluates the request:

~~~
evaluatedElement, proof = server_context.Evaluate(key, config.public_key, blindedElement, info)
~~~

The server sends both `evaluatedElement` and `proof` to the client. These are concatenated together. As the length of both is fixed, there is no ambiguity in parsing the result.

The client then completes issuance as follows:

~~~
output = client_context.Finalize(nonce, blind, evaluatedElement, info):
~~~

This procedure may fail with an error (`VerifyError` or `DeserializeError`), in which case the issuance is said to have failed. The output of the issuance protocol is the concatenation of `nonce` and `output`, denoted as `token`:

~~~
struct {
   uint8 nonce[32];
   uint8 output[Nh]; // Nh is as defined in {{I-D.irtf-cfrg-voprf}}
} Token;
~~~

In the redemption phase, the client sends the Token to the server to verify locally. In particular, the server verifies the Token as follows:

~~~
valid = server_context.VerifyFinalize(key, token.nonce, token.output, info)
~~~

Redemption is considered successful if `valid` is true.
```